### PR TITLE
Breaking API changes for v3

### PR DESCRIPTION
**Pull request by christopheranderson**
Opening PR for feedback.

This PR has the first wave of the breaking API changes in it.
- Changes all response types to be classes derived from ResourceResponse
- Adds getters for important header values
- Removes nextItem/current from queryIterator
- Changes response type for queryIterator to always be a page of data
- Fixes a bunch of circular dependencies (does not add any testing for them, though)
- Renames IHeaders to CosmosHeaders
- Deletes ruPerMin tests
- Deletes some helper tests
- Changes status codes to object with "as #" names
- Engineering: Adds a common js compile target and vs-code uses that for debugging
- Updates ts-node dependency

Todo:
- [ ] Change all the Options on arguments to be resource/operation specific
- [ ] Decide if we need partition key specified for everything (needs debate)
- [ ] ???

I'd prefer to do one big breaking change PR, so need to go through our list still to see if I've missed anything. Feel free to modify the PR todo list if you see anything.

### CODE REVIEW

**Comment by christopheranderson** [comment]
FYI - a bunch of files got touched when I was trying to remove circular dependencies that started to break my build. We should think about having rollup actually throw on new circular dependencies being added. I had to manually scan through the log output, but I got rid of them all.

File: src/client/Container/Container.ts [hunk 1]
```diff
@@ -109,7 +141,7 @@ export class Container {
    * Use `.userDefinedFunctions` for creating new user defined functions, or querying/reading all user defined functions.
    * @param id The id of the {@link UserDefinedFunction}.
    */
-  public userDefinedFunction(id: string): UserDefinedFunction {
+  protected __userDefinedFunction(id: string): UserDefinedFunction {
```
Comment (southpolesteve) [inline src/client/Container/Container.ts]: Why adding the double underscore prefix?

File: src/request/FeedOptions.ts [hunk 2]
```diff
@@ -1,47 +1,47 @@
-import { IHeaders } from "../index";
+import { CosmosHeaders } from "../index";
 
 /**
  * The feed options and query methods.
  */
-export interface FeedOptions {
+export class FeedOptions {
```
Comment (southpolesteve) [inline src/request/FeedOptions.ts]: Why make this a class?

File: tsconfig.commonjs.json [hunk 3]
```diff
@@ -0,0 +1,20 @@
+{
```
Comment (southpolesteve) [inline tsconfig.commonjs.json]:
```suggestion
{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "module": "commonjs"
  }
}
```

File: src/queryIterator.ts [hunk 4]
```diff
@@ -135,27 +125,29 @@ export class QueryIterator<T> {
   /**
    * Retrieve all the elements of the feed and pass them as an array to a function
    */
-  public async toArray(): Promise<Response<T[]>> {
+
+  public async toArray(): Promise<FeedResponse<T>> {
```
Comment (southpolesteve) [inline src/queryIterator.ts]: Do we need a new special type for the `toArray` repsonse that remembers all the headers? I think the suffers from the same issue we have in v2 where toArray only has the last header values

File: src/client/Container/Container.ts (hunk 1 above)
Comment (christopheranderson) [reply src/client/Container/Container.ts]: single underscore is yelled at by tslint and I want a different name than the old type so we can still build our mocked types using the old api surface area.

File: src/queryIterator.ts (hunk 4 above)
Comment (christopheranderson) [reply src/queryIterator.ts]: This is a good point. I'll think on it.

File: src/request/FeedOptions.ts (hunk 2 above)
Comment (christopheranderson) [reply src/request/FeedOptions.ts]: I think this is a mistake. I'm not sure why I did this.

File: src/queryIterator.ts (hunk 4 above)
Comment (christopheranderson) [reply src/queryIterator.ts]: We're not saving any headers anymore, so 1 FeedResponse is good enough.

**Comment by christopheranderson** [comment]
Created #248 to track fixing up samples after these changes.

**Pull request merged**